src/forms/Typeahead.jsx

```diff
@@ -12,6 +12,54 @@ export const TA_ITEM_CLASSNAME = 'typeahead-item';
  * @module Typeahead
  */
 class Typeahead extends React.PureComponent {
+	constructor(props) {
+		super(props);
+		this.onItemMouseEnter = this.onItemMouseEnter.bind(this);
+		this.stateReducer = this.stateReducer.bind(this);
+
+		this.state = {
+			highlightedIndex: -1
+		};
+	}
+
+	onItemMouseEnter (i) {
+		this.setState({highlightedIndex: i});
+	}
+
+	stateReducer(state, changes) {
+		let highlightedIndexState;
+
+		switch (changes.type) {
+			case Downshift.stateChangeTypes.keyDownArrowDown:
+				highlightedIndexState =
+					parseInt(this.state.highlightedIndex + 1) >= this.props.items.length
+						?
+							0
+						:
+							parseInt(this.state.highlightedIndex+1);
+					this.setState({
+						highlightedIndex: highlightedIndexState
+					});
+				break;
+			case Downshift.stateChangeTypes.keyDownArrowUp:
+				highlightedIndexState =
+					this.state.highlightedIndex < 1
+						?
+							parseInt(this.props.items.length-1)
+						:
+							parseInt(this.state.highlightedIndex-1);
+					this.setState({
```

Comment:
The internal `setState` can also happen outside the switch statement. The `switch` would then only have the responsibility of updating the value of `highlightedIndexState`

Reply:
I tried that, it doesn't work for some reason